src/main/java/pro/belbix/ethparser/web3/harvest/HardWorkCalculator.java

```diff
@@ -0,0 +1,119 @@
+package pro.belbix.ethparser.web3.harvest;
+
+import org.springframework.stereotype.Service;
+import pro.belbix.ethparser.dto.HarvestDTO;
+import pro.belbix.ethparser.repositories.HardWorkRepository;
+import pro.belbix.ethparser.repositories.HarvestRepository;
+import pro.belbix.ethparser.web3.EthBlockService;
+import pro.belbix.ethparser.web3.prices.PriceProvider;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class HardWorkCalculator {
+
+    private final double ETH_ESTIMATE = 0.1;
+    private final PriceProvider priceProvider;
+    private final HarvestRepository harvestRepository;
+    private final HardWorkRepository hardworkRepository;
+    private final EthBlockService ethBlockService;
+
+    public HardWorkCalculator(PriceProvider priceProvider, HarvestRepository harvestRepository, HardWorkRepository hardWorkRepository, EthBlockService ethBlockService) {
+        this.priceProvider = priceProvider;
+        this.harvestRepository = harvestRepository;
+        this.hardworkRepository = hardWorkRepository;
+        this.ethBlockService = ethBlockService;
+    }
+
+    public double calculateTotalHardWorksFeeByOwner(String ownerAddress) {
+        long lastBlock = ethBlockService.getLastBlock();
+        long lastBlockDate = ethBlockService.getTimestampSecForBlock(null, lastBlock);
+        
+        List<HarvestDTO> harvests = harvestRepository.fetchAllByOwner(ownerAddress, 0, lastBlockDate);
+        List<HarvestDTO> sortedHarvests = harvests
+            .stream()
+            .sorted(Comparator.comparingLong(HarvestDTO::getBlockDate))
+            .collect(Collectors.toList());
+        
+        HashMap<String, ArrayList<ArrayList<Long>>> blockRangeByVault = new HashMap<>();
+        for (HarvestDTO harvest : sortedHarvests) {
+            String vault = harvest.getVault();
+            double balance = harvest.getOwnerBalance();
+
+            if (!blockRangeByVault.containsKey(vault)) {
+                if (balance == 0) {
+                    continue;
+                }
+                ArrayList<ArrayList<Long>> blockPeriods = new ArrayList<>();
```

Comment:
you can use ArrayList<Tuple<Long, Long>> and keep only start-end blocks
it will be more clear
or a primitive array with size 2

Reply:
I updated it to use a fixed primitive array with size 2, because the existing tuple library won't support mutation, which is needed when updating an initialized block period with an end date.